src/main/java/ch/njol/skript/SkriptEventHandler.java

```diff
@@ -79,74 +81,73 @@ public PriorityListener(EventPriority priority) {
 	}
 
 	/**
-	 * A list tracking what Triggers are paired with what Events.
+	 * A Multimap tracking what Triggers are paired with what Events.
+	 * Each Event effectively maps to an ArrayList of Triggers.
 	 */
-	private static final List<NonNullPair<Class<? extends Event>, Trigger>> triggers = new ArrayList<>();
+	private static final Multimap<Class<? extends Event>, Trigger> triggers = ArrayListMultimap.create();
 
 	/**
-	 * A utility method to get all Triggers paired with the provided Event class.
+	 * A utility method to get all Triggers registered under the provided Event class.
 	 * @param event The event to find pairs from.
-	 * @return An iterator containing all Triggers paired with the provided Event class.
+	 * @return A List containing all Triggers registered under the provided Event class.
 	 */
-	private static Iterator<Trigger> getTriggers(Class<? extends Event> event) {
+	private static List<Trigger> getTriggers(Class<? extends Event> event) {
 		HandlerList eventHandlerList = getHandlerList(event);
 		assert eventHandlerList != null; // It had one at some point so this should remain true
-		return new ArrayList<>(triggers).stream()
-			.filter(pair -> pair.getFirst().isAssignableFrom(event) && eventHandlerList == getHandlerList(pair.getFirst()))
-			.map(NonNullPair::getSecond)
-			.iterator();
+		return triggers.asMap().entrySet().stream()
+				.filter(entry -> entry.getKey().isAssignableFrom(event) && getHandlerList(entry.getKey()) == eventHandlerList)
+				.flatMap(entry -> entry.getValue().stream())
+				.collect(Collectors.toList()); // forces evaluation now and prevents us from having to call getTriggers again if very high logging is enabled
 	}
 
 	/**
 	 * This method is used for validating that the provided Event may be handled by Skript.
 	 * If validation is successful, all Triggers associated with the provided Event are executed.
 	 * A Trigger will only be executed if its priority matches the provided EventPriority.
-	 * @param e The Event to check.
+	 * @param event The Event to check.
 	 * @param priority The priority of the Event.
 	 */
-	private static void check(Event e, EventPriority priority) {
-		Iterator<Trigger> ts = getTriggers(e.getClass());
-		if (!ts.hasNext())
+	private static void check(Event event, EventPriority priority) {
+		List<Trigger> triggers = getTriggers(event.getClass());
+		if (triggers.isEmpty())
```

Comment:
I don't think this check is needed. At least, the key should be removed from the list in the unregister method if all triggers have been unregistered.